HGCTPG Stage1 emulator

portalesHEP opened the pull request:
Including the setup for standalone S1 emulator tests. Right now it is independent of the S2 one, so it could technically be kept entirely separated by creating different folders/branches, though the config/input parser that I included could be used for S2 as-is as well if it seems useful.

 Some things worth noting:

- the `getCMSSWCode_S1.sh` script retrieves code from my CMSSW fork. This code is mostly independent of the S2 code, with the exception of `L1Trigger/L1THGCal/interface/backend_emulator/HGCalTriggerCell_SA.h`, which I had to modify slightly to be used inside CMSSW for S1 (where I needed to propagate an index variable, as done e.g., for clusters). Due to the difference, it is not possible right now to retrieve both the S1 and S2 dependencies simultaneously. I'm not sure yet what the best way to deal with this would be.
- As stated in the README, I stored some temporary input/config files on eos, which is probably not ideal, and I'll also think about a better way to access them (git-lfs?)
- The necessary configurable inputs are now read from a JSON file through [nlohmann_json](https://github.com/nlohmann/json).  The package is available in CMSSW but may need to be installed for standalone testing. Instructions to do so in a conda environment are included in the README.

portalesHEP committed:
initial commit for S1

portalesHEP committed:
more consistent git setup scripts

portalesHEP committed:
some cleaning + include missing python scripts

portalesHEP committed:
removing useless header

portalesHEP committed:
cleaning

portalesHEP committed:
cleaning

portalesHEP committed:
Modified config parser to read JSON files

portalesHEP committed:
changes missing from previous commit

portalesHEP committed:
Replace json handling package for nlohmann_json (used in CMSSW); added setup instructions; changed format of input TC list to json; passing function arguments by reference

EmyrClement commented:
@portalesHEP I was just taking a look through, and see you've added another commit, which answer some of the comments I was making, so I'll pause the review - are there more coming soon, or should I carry on?

portalesHEP commented:
Hi @EmyrClement, thanks for having a look. I would still like to change the format of the TCmap from csv to json for consistency (and hopefully for better readability). I might also change the structure of `HGCTPGEmulatorTester_S1.cc`, possibly to allow for tests on both stage 1 and 2 from the same executable. This is not on top of my priority list at the moment though, and could be done in a future PR, so you can probably proceed with the review.

EmyrClement commented on HGCTPGEmulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,37 @@
+#include "L1Trigger/L1THGCal/interface/backend_emulator/HGCalTriggerCell_SA.h"
+
+#include <nlohmann/json.hpp>
+using json = nlohmann::json;
```
Using these json tools for consistency with CMSSW was going to be my first comment.  We should make sure we apply setup instructions for getting the tools.

portalesHEP replied:
I included a short paragraph to the README for the setup in a conda environment. Alternatively, if conda is not a good option for some reason, the package's README on github ([here](https://github.com/nlohmann/json)) goes through the setup options quite extensively.

EmyrClement commented on HGCTPGEmulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,37 @@
+  ~SAConfigParser() {}
+
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > parseTClist(const std::string& theInputFile) const;
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > parseTClist(const std::string& theInputFile, const std::string& theTCMap) const;
```
Should we separate the tools for reading in data from those that read the config parameters?  In the end, I hope the parsing of the input data will be done by the `DemonstratorTools`, so separating the data/config parsing tools from the start will help us switch in the future.

portalesHEP replied:
Indeed that would make sense, I'll look into it.

portalesHEP committed:
fetching correct branch for CMSSW dependancies + minor fixes

portalesHEP commented:
@EmyrClement sorry I spoke a bit too soon: I realised there was a compatibility issue with the ongoing PR in [hgc-tpg/cmssw](https://github.com/hgc-tpg/cmssw/pull/4): the shell script to retrieve the cmssw dependancies still pointed to an older branch, and the up-to-date version uses slightly different names. I just commited the fix, and there shouldn't be any more updates for now.

EmyrClement commented on python/extract_mappings.py:
```diff
@@ -0,0 +1,306 @@
+#!/usr/bin/env python3
```
Is this file used anywhere?

portalesHEP replied:
No, it was used by mistake at first, in place of the `load_mappings.py` script, and I forgot to remove it. I'll do it in my next commit.

EmyrClement commented on README.md:
```diff
@@ -0,0 +1,46 @@
+
+- `inputs/TClist.txt`: list generated from `ModXXXX_stimuli.txt` files, containing input TCs per module, as used for S1 firwmare testing. 
+
+  For these, additional information on the mapping of FPGA/modules/TCs in <img src="https://render.githubusercontent.com/render/math?math=(r/z,\phi)"> is needed. These are `.csv` files, generated with `python/generateTCMap.py`.
```
Do the files in the python directory belong in this repo, or are they pulled in from somewhere else?

portalesHEP replied:
`load_mappings.py` is taken from the S1 firmware generator repository. The other two scripts are specific to `HGCTPGEmulator`, but depends on the former script. However all these scripts will in principle not be useful anymore once we settle on the parameter extraction in CMSSW.

EmyrClement commented on README.md:
```diff
@@ -0,0 +1,46 @@
+
+  To compile all the components required for tests, simply call `make` from this directory. The executable, `HGCTPGEmulatorTester_S1` will be located in `test`, and some basic configurations to run it are available in `test/run.sh`.
+
+  Right now, `HGCTPGEmulatorTester_S1` expects either two or three input parameters: either a list of TCs in the "S2 format" and a configuration file (example here: `inputs/config_S1SortingTruncation_v0_20211129.cfg`), or a list of TCs in the "S1 format", with the same configuration file, in addition to the `.csv` TC map generated with `python/generateTCMap.py`.
```
Could you provide a list of commands to run the setup, copying any inputs to the local area (assuming access to `/eos`), and run the emulation?  It would help me checkout the code and run it, and would probably be useful for CI in the future.

portalesHEP replied:
Here's how to run the tests (I'll add that to the README in the next commit as well): 
```
# get HGCTPGEmulator & CMSSW dependancies
mkdir HGCTPGEmulator
cd HGCTPGEmulator
git init
mv .git/ .githgctpg
git --git-dir=.githgctpg remote add -t HGCTPGEmulator_S1 S1Emulator git@github.com:portalesHEP/HGCTPGEmulator.git
git --git-dir=.githgctpg pull S1Emulator HGCTPGEmulator_S1
source getCMSSWCode_S1.sh

# Create conda env with required packages
conda create --name myEnv
conda activate myEnv
conda install -c conda-forge nlohmann_json pkg-config

# Compile & run tester
make
cd test
source run.sh
```
About the files on EOS, sorry I didn't realize these weren't accessible. l copied them on afs (here: `/afs/cern.ch/user/l/lportale/public/HGCTPG/`) as a temporary workaround, but I'll see about changing the authorizations.

EmyrClement reviewed (commented):
I made some very general comments for now.  Regarding your comment on having both the S1 and S2 code in the same repository, perhaps for now we separate the two into different directories.  We could discuss this at the TPG meeting later today.

portalesHEP commented:
Here's how to run the tests (I'll add that to the README in the next commit as well): 
```
# get HGCTPGEmulator & CMSSW dependancies
mkdir HGCTPGEmulator
cd HGCTPGEmulator
git init
mv .git/ .githgctpg
git --git-dir=.githgctpg remote add -t HGCTPGEmulator_S1 S1Emulator git@github.com:portalesHEP/HGCTPGEmulator.git
git --git-dir=.githgctpg pull S1Emulator HGCTPGEmulator_S1
source getCMSSWCode_S1.sh

# Create conda env with required packages
conda create --name myEnv
conda activate newEnv
conda install -c conda-forge nlohmann_json pkg-config

# Compile & run tester
make
cd test
source run.sh
```

portalesHEP committed:
updated readme with setup/run instructions; removed unused python script; updated git branch from which CMSSW code is fetched (no functional difference, but now consistent with ongoing PR in CMSSW)

portalesHEP committed:
Revamping the emulator tester structure

portalesHEP committed:
Merge pull request #1 from portalesHEP/master

Revamping the stage1 emulator tester structure

portalesHEP commented:
The last commit (e511de7) contains significant modifications that are listed here: 
- Moved the whole Stage1 emulator tests in its own directory `HGCTPGStage1Emulator`
- Improved code structure: 
    - `test/HGCTPGStage1EmulatorTester.cc`: moved tests outside of main function, now all done in utility class `HGCTPGEmulatorTester`
    - `bin/HGCTPGStage1EmulatorTester.exe` generated at compilation. Takes as arguments (in that order) a list of input TCs (`json` format), a config file (`json`, generated with CMSSW analyzer [HGCalBackendStage1ParameterExtractor.cc](https://github.com/hgc-tpg/cmssw/pull/11/files/57274bab0ed8794fd69c79b592aeb39b64584544#diff-8ec5a3a15d5d69de0952c2e7d5d163737cf019284952f8aa02aadca2243723e9)), and a path to a directory where the test outputs will be stored.
    - The TC list and Config reader classes have been split according to comments: `SAConfigParser.cc` and `SATCParser.cc`.
    - The `nlohmann_json` header file is now included to avoid having to ensure it is installed manually. (could be fetched from git on setup instead to keep it updated if that sounds better).
- Changed tester output format: now returning tables of hexadecimal values (TC address + energy) per (roz,phi) bin matching the output format of Stage 1 firmware. NOTE: the module hash is currently saved in place of the TC addresses; this will have to be changed to match the firmware definition

portalesHEP committed:
updated python script for TC list generation

portalesHEP committed:
Merge pull request #2 from portalesHEP/master

updated python script for TC list generation

portalesHEP committed:
fix output

portalesHEP committed:
TC list generation script now takes input arguments

jbsauvan commented on HGCTPGStage1Emulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,47 @@
+  // getters
+  l1thgcfirmware::Stage1TruncationConfig getCfg() const {return theCfg_;}
+  std::map< std::pair<unsigned,unsigned>, std::pair<double,double> > getTCmap() const {return theTCMap_;}
+  const double magic_number(){return rozphi_scale_;}
```
- `magic_number` is not very explicit.
- If `rozphi_scale_` is needed outside this class,  could it be public?

portalesHEP replied:
removed all ```magic_number``` and ```rozphi_scale_``` occurences, that were only used before bin numbers were introduced in the global json configuration.

jbsauvan commented on HGCTPGStage1Emulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,47 @@
+
+private:
+  l1thgcfirmware::Stage1TruncationConfig parseCfg(const std::string& theCfgFile) const;
+  std::map< std::pair<unsigned,unsigned>, std::pair<double,double> > setTCmap(const std::string& theTCMap) const;
```
`setTCmap` seems to be only used to set  `theTCMap_`, so it is a bit strange to make it return the full map instead of directly accessing/filling `theTCMap_`. 
Directly filling `theTCMap_` would avoid unnecessary copy

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,47 @@
+
+  // getters
+  l1thgcfirmware::Stage1TruncationConfig getCfg() const {return theCfg_;}
+  std::map< std::pair<unsigned,unsigned>, std::pair<double,double> > getTCmap() const {return theTCMap_;}
```
Couldn't it return a const reference to avoid the copy?

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,47 @@
+  ~SAConfigParser() {}
+
+  // getters
+  l1thgcfirmware::Stage1TruncationConfig getCfg() const {return theCfg_;}
```
Couldn't it return a const reference to avoid the copy?

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/interface/SATCParser.h:
```diff
@@ -0,0 +1,49 @@
+  ~SATCParser() {}
+
+
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > TClist() const {return theTClist_;}
```
Couldn't it return a const reference to avoid the copy?

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/interface/SATCParser.h:
```diff
@@ -0,0 +1,49 @@
+
+
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > TClist() const {return theTClist_;}
+  const double magic_number(){return rozphi_scale_;}
```
- Same as before
- Also, may be not the best to duplicate the same constant in multiple places

portalesHEP replied:
removed all ```magic_number``` and ```rozphi_scale_``` occurences, that were only used before bin numbers were introduced in the global json configuration.

jbsauvan commented on HGCTPGStage1Emulator/interface/SATCParser.h:
```diff
@@ -0,0 +1,49 @@
+
+private:
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > parseTClist_s2(const std::string& theInputFile) const;
+  std::vector< std::vector <l1thgcfirmware::HGCalTriggerCell> > parseTClist(const std::string& theInputFile, const TCMap& theTCMap) const;
```
It seems to be only used to fill `theTClist_` so it would be better to directly fill `theTClist_` to avoid unnecessary copy

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/interface/SAConfigParser.h:
```diff
@@ -0,0 +1,47 @@
+  const double magic_number(){return rozphi_scale_;}
+
+private:
+  l1thgcfirmware::Stage1TruncationConfig parseCfg(const std::string& theCfgFile) const;
```
It seems to be only used to set `theCfg_`, so  it would be better to directly fill `theCfg_` to avoid unnecessary copies

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+
+  // Get list of TCs to be processed
+  std::ifstream theFile(theTCMap);
+  if (!theFile.is_open()) throw std::runtime_error("Could not open Mod_stimuli list.");
```
Is this exception catched somewhere?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+  std::map< std::pair<unsigned,unsigned>, std::pair<double,double> > TCmap_out;
+  for (auto& module: theCfgJSON["Module"]) {
+
+    unsigned moduleHash = std::stoul((std::string)module.at("hash"));
```
The string is converted to `unsigned long` but then the `moduleHash` is downgraded to an `unsigned`. May it cause issues?

portalesHEP replied:
the module hash is stored as a ```uint32_t``` before being stored in the json config, and should in any case never be more than 14 bits by construction, so I'm not expecting issues here.

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+
+  // Read json file into string
+  std::ifstream theJSONCfg(theCfgFile);
+  if (!theJSONCfg.is_open()) throw std::runtime_error("Could not open config file.");
```
Is this exception catched somewhere?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+  } catch (const json::exception& e) {
+    throw std::runtime_error("Could not read required parameter: doTruncation");
+  }
+  */
```
Commented code to be removed

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+    rozMin = (double)theCfg.at("rozMin") * rozphi_scale_;
+    rozMax = (double)theCfg.at("rozMax") * rozphi_scale_;
+  } catch (const json::exception& e) {
+    throw std::runtime_error("Could not read required parameter: rozMin,rozMax");
```
Catched exception?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+  try {
+    rozBins = theCfg.at("rozBins");
+  } catch (const json::exception& e) {
+    throw std::runtime_error("Could not read required parameter: rozBins");
```
Catched exception?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+  double rozMax = 0.;
+  try {
+    rozMin = (double)theCfg.at("rozMin") * rozphi_scale_;
+    rozMax = (double)theCfg.at("rozMax") * rozphi_scale_;
```
Why is this `rozphi_scale_` scale factor needed?

portalesHEP replied:
Removed all magic_number and rozphi_scale_ occurences, that were only used before bin numbers were introduced in the global json configuration.it was needed to deal with the fact that the HGCalTriggerCell class works with `unsigned int`. The parameter value was the one used for stage 2, and held no proper meaning for stage 1.

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+  std::vector<double> phiSectorEdges;
+  try {
+    for (const double& theEdge: theCfg.at("phiSectorEdges"))
+      phiSectorEdges.push_back(theEdge * rozphi_scale_);
```
Why is this rozphi_scale_ scale factor needed?

portalesHEP replied:
Removed all magic_number and rozphi_scale_ occurences, that were only used before bin numbers were introduced in the global json configuration.it was needed to deal with the fact that the HGCalTriggerCell class works with unsigned int. The parameter value was the one used for stage 2, and held no proper meaning for stage 1.

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+    for (const double& theEdge: theCfg.at("phiSectorEdges"))
+      phiSectorEdges.push_back(theEdge * rozphi_scale_);
+  } catch (const json::exception& e) {
+    throw std::runtime_error("Could not read required parameter: phiSectorEdge");
```
catched exception?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+    phi = 2.*M_PI - phi + (4. * M_PI / 3.);
+
+  return phi;
+}
```
Do we need to rotate phi in the config parser? In principle the config generated by CMSSW should use rotated coordinates?

portalesHEP replied:
Indeed this is not needed anymore. This function has been removed

jbsauvan commented on HGCTPGStage1Emulator/src/SATCParser.cc:
```diff
@@ -0,0 +1,140 @@
+
+  // Get list of TCs to be processed
+  std::ifstream theFile(theInputFile);
+  if (!theFile.is_open()) throw std::runtime_error("Could not open Mod_stimuli list.");
```
Catched exception?

portalesHEP replied:
Exceptions are now handled in constructor

jbsauvan commented on HGCTPGStage1Emulator/src/SAConfigParser.cc:
```diff
@@ -0,0 +1,126 @@
+
+  // loop on events
+  std::map< std::pair<unsigned,unsigned>, std::pair<double,double> > TCmap_out;
+  for (auto& module: theCfgJSON["Module"]) {
```
Can it be `const`?

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/src/SATCParser.cc:
```diff
@@ -0,0 +1,140 @@
+  theFile >> theTCjson;
+  int evt=0;
+  // loop on events
+  for (auto& event: theTCjson.at("event")) {
```
Can it be `const`?

portalesHEP replied:
const qualifier causes compilation issues here. leaving as is for now, but added a comment explaining the situation

jbsauvan commented on HGCTPGStage1Emulator/src/SATCParser.cc:
```diff
@@ -0,0 +1,140 @@
+      for (const std::string& tc: moduleTcs.at("tc")) {
+
+	// get rz/phi values from map
+	std::pair<unsigned,unsigned> modHash_tcid{moduleHash,tcid};
```
Here and below: indent issue (mixed spaces and tabs)

jbsauvan commented on HGCTPGStage1Emulator/src/SATCParser.cc:
```diff
@@ -0,0 +1,140 @@
+	  
+	  tmpTC.setIndex(moduleHash); // 
+	  // fill event's TC vector
+	  TCs_tmp.push_back(tmpTC);
```
Would be better to construct the TC in-place, using `emplace_back`, to avoid the copy

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/src/SATCParser.cc:
```diff
@@ -0,0 +1,140 @@
+
+    std::cout << std::endl;
+    evt++;
+    TCs_out.push_back(TCs_tmp);
```
Could directly fill `TCs_out` instead of copying the full vectors of TCs

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+  int dumpTCs_fw(const HGCalTriggerCellsPerBx&, const std::string);
+
+  //getter
+  HGCalTriggerCellsPerBx getTCs() {return allTCs_;}
```
Should return a `const` reference to avoid copy

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+
+public:
+
+  HGCTPGEmulatorTester(const std::string, const std::string);
```
Better to use reference: `const std::string&`

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+int HGCTPGEmulatorTester::runAlgo(const HGCalTriggerCellsPerBx& tcPerBx, HGCalTriggerCellsPerBx& sortedTruncatedTCsPerBx) {
+  
+  unsigned int counter = 0;
+  for (const auto TCs : tcPerBx) {
```
`const auto&` to avoid copies

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+  unsigned int counter = 0;
+  for (const auto TCs : tcPerBx) {
+    // Run the algorithm
+    HGCalTriggerCells sortedTruncatedTCs;
+    unsigned error_code = stage1Algo_.run(TCs, stage1Config_, sortedTruncatedTCs);
+    sortedTruncatedTCsPerBx.push_back(sortedTruncatedTCs);
```
Couldn't the `sortedTruncatedTCsPerBx` be directly filled instead of copying things around?

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+  std::ofstream outputStream;
+  outputStream.open(outputFileName);
+  unsigned int counter = 0;
+  for (const auto TCs: tcPerBx) {
```
Better to use reference: `const auto&`

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+  // Create TC vectors per (roz,phi) bin
+  unsigned event=0;
+  std::unordered_map< unsigned, std::vector<std::pair<unsigned, HGCalTriggerCell> > > TCandEvtperBin;
+  for (const auto TCs: tcPerBx) {
```
Better to use reference: `const auto&`

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+      packed_bin |= ((rzbin & mask_roz_) << offset_roz_);
+      packed_bin |= (phibin & mask_phi_);
+      
+      TCandEvtperBin[packed_bin].push_back(std::make_pair(event,tc));
```
Better to use `emplace_back` to construct the pair in-place and avoid the copy

portalesHEP replied:
done

jbsauvan commented on HGCTPGStage1Emulator/test/HGCTPGStage1EmulatorTester.cc:
```diff
@@ -0,0 +1,235 @@
+    for (const auto& evtAndTc: bin_tcs.second) {
+      unsigned theEvent = evtAndTc.first;
+      if(theEvent!=ievt){
+	outputStream << std::endl;
+	ievt++;
```
Indent issue

jbsauvan commented on Makefile:
```diff
@@ -0,0 +1,32 @@
+### HGCTPGEmulatorTester_S1 Makefile
```
Shouldn't this Makefile be in the Stage 1 subdirectory?

portalesHEP replied:
the Makefile also takes care of compiling the classes from CMSSW (in `L1Trigger` directory), so it seemed to be an adequate location. This could still be changed though.

jbsauvan replied:
I guess then it should also at some point compile the Stage 2 code. So that's ok if it is the intent. Otherwise the file should probably have a name that explicitly tells it is for Stage 1 only.

jbsauvan commented on README.md:
````diff
@@ -0,0 +1,75 @@
+
+## Dependencies
+
+The testing framework uses the `nlohmann_json` library, which must be installed and accessible. If not, it can be easily set, e.g. in a conda environment:
+
+```
+conda create --name newEnv
+conda activate newEnv
+conda install -c conda-forge nlohmann_json pkg-config
+```
+
+`pkg-config` is also needed inside the conda environment, in order to automatically pick up the package location inside the Makefile.
````
Is it still needed since the json library is now included?

portalesHEP replied:
updated README

jbsauvan commented on scripts/generateTCMap.py:
```diff
@@ -0,0 +1,26 @@
+import pandas as pd
```
Is this file still needed?

portalesHEP replied:
it is not, now removed

jbsauvan commented on scripts/load_mappings.py:
```diff
@@ -0,0 +1,332 @@
+
```
Is this file still needed?

portalesHEP replied:
it is not, now removed

portalesHEP committed:
fetching CMSSW through http in setup script

portalesHEP committed:
automatic evaluation of number of BX in TC list generation script

portalesHEP committed:
new output TCid definition; storing module hash + tc id (should be same as in fw)

portalesHEP committed:
modifying output formatting

portalesHEP committed:
reading tc roz/phi bin directly from cfg

portalesHEP committed:
changed input dump format

portalesHEP committed:
changing dump format for input tcs

portalesHEP committed:
changing dump format when dumpAddresses==false

portalesHEP committed:
removing unused instances of rozphi_scale; setTCmap now fills theTCMap_ directly; some mild additional cleaning

portalesHEP committed:
returning const references with getters

portalesHEP committed:
parseTClist now directly filling theTClist_

portalesHEP committed:
parseCfg now filling theCfg_ on the fly

portalesHEP committed:
handling exceptions in SAConfigParser

portalesHEP committed:
removing rozphi_scale_, not needed anymore as fetching roz/phi bins directly from config

portalesHEP committed:
removing unused phi rotation function

portalesHEP committed:
in parseTClist: push_back -> emplace_back

portalesHEP committed:
last comments implemented in HGCTPGStage1EmulatorTester.cc; removing deprecated python scripts

portalesHEP committed:
updated README

portalesHEP committed:
catching exception in SATCParser; adding comment on non-critical issue with nlohmann_json

portalesHEP committed:
adding missing const qualifiers

portalesHEP committed:
returning reference with getTCs()

portalesHEP committed:
missing reference qualifier in loop initialization

portalesHEP committed:
fixes to match changes in backend_emulator/HGCalTriggerCell_SA.h

portalesHEP committed:
switching to C++17

portalesHEP committed:
now setting up CMSSW environment for C++ version consistency

portalesHEP committed:
bugfix in CMSSW setup
